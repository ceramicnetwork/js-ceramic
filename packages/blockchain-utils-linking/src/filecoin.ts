import { AuthProvider } from "./auth-provider";
import { AccountID } from "caip";
import { getConsentMessage, LinkProof } from "./util";
import type { MessageParams } from "@zondax/filecoin-signing-tools";
import * as uint8arrays from "uint8arrays";

export class FilecoinAuthProvider implements AuthProvider {
  readonly isAuthProvider = true;

  constructor(
    private readonly provider: any,
    private readonly address: string
  ) {}

  async accountId(): Promise<AccountID> {
    const prefix = this.address[0];
    const chainId = `fil:${prefix}`;
    return new AccountID({ address: this.address, chainId });
  }

  async authenticate(message: string): Promise<string> {
    const payload = asTransaction(this.address, JSON.stringify(message));
    const signatureResponse = await this.provider.sign(this.address, payload);
    return signatureResponse.Signature.Data;
  }

  async createLink(did: string): Promise<LinkProof> {
    const { message, timestamp } = getConsentMessage(did, true);
    const payload = asTransaction(this.address, message);
    const signatureResponse = await this.provider.sign(this.address, payload);
    const accountId = await this.accountId();
    return {
      version: 2,
      type: "eoa-tx",
      message: message,
      signature: signatureResponse.Signature.Data,
      account: accountId.toString(),
      timestamp: timestamp,
    };
  }

  withAddress(address: string): AuthProvider {
    return new FilecoinAuthProvider(this.provider, address);
  }
}

export function asTransaction(address: string, message: string): MessageParams {
  const messageParams = uint8arrays.toString(
    uint8arrays.fromString(message),
    "base64"
  );
  return {
    From: address,
    To: address,
    Value: "0",
    Method: 0,
    GasPrice: "1",
    GasLimit: 1000,
    Nonce: 0,
    Params: messageParams,
    GasFeeCap: "1",
    GasPremium: "1",
  };
}
