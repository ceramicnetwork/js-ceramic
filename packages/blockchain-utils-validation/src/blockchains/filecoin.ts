import { BlockchainHandler } from "../blockchain-handler";
import { AccountID } from "caip";
import * as linking from "@ceramicnetwork/blockchain-utils-linking";

const namespace = "fil";

export async function validateLink(
  proof: linking.LinkProof
): Promise<linking.LinkProof | null> {
  const signingTools = await import("@zondax/filecoin-signing-tools")
  const account = new AccountID(proof.account);
  const payload = linking.filecoin.asTransaction(
    account.address,
    proof.message
  );
  const transaction = signingTools.transactionSerialize(payload);
  try {
    const recover = signingTools.verifySignature(proof.signature, transaction);
    if (recover) {
      return proof;
    } else {
      return null;
    }
  } catch {
    return null;
  }
}

const Handler: BlockchainHandler = {
  namespace,
  validateLink,
};

export default Handler;
