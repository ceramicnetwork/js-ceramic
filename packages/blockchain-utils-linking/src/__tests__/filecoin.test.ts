import { LocalManagedProvider } from "@glif/local-managed-provider";
import { Network } from "@glif/filecoin-address";
import { FilecoinAuthProvider } from "../filecoin";

const did = "did:3:bafysdfwefwe";
const testnetPrivateKey =
  "7b2254797065223a22736563703235366b31222c22507269766174654b6579223a2257587362654d5176487a366f5668344b637262633045642b31362b3150766a6a504f3753514931355031343d227d";
const mainnetPrivateKey =
  "7b2254797065223a22736563703235366b31222c22507269766174654b6579223a2257587362654d5176487a366f5668344b637262633045642b31362b3150766a6a554f3753514931355031343d227d";
const blsPrivateKey =
  "7b2254797065223a22626c73222c22507269766174654b6579223a226e586841424f4163796856504b48326155596261796f4475752f4c6f32515a2b6662622f6f736a2f34456f3d227d";
const testnetProvider = new LocalManagedProvider(
  testnetPrivateKey,
  Network.TEST
);
const mainnetProvider = new LocalManagedProvider(
  mainnetPrivateKey,
  Network.MAIN
);
const blsMainnetProvider = new LocalManagedProvider(
  blsPrivateKey,
  Network.MAIN
);

beforeAll(() => {
  global.Date.now = jest.fn().mockImplementation(() => 666);
});

afterAll(() => {
  jest.clearAllMocks();
});

describe("accountId", () => {
  test("handle testnet", async () => {
    const addresses = await testnetProvider.getAccounts();
    const authProvider = new FilecoinAuthProvider(
      testnetProvider,
      addresses[0]
    );
    const accountId = await authProvider.accountId();
    expect(accountId).toMatchSnapshot();
  });
  test("handle mainnet", async () => {
    const addresses = await mainnetProvider.getAccounts();
    const authProvider = new FilecoinAuthProvider(
      mainnetProvider,
      addresses[0]
    );
    const accountId = await authProvider.accountId();
    expect(accountId).toMatchSnapshot();
  });
  test("handle bls mainnet", async () => {
    const addresses = await blsMainnetProvider.getAccounts();
    const authProvider = new FilecoinAuthProvider(
      blsMainnetProvider,
      addresses[0]
    );
    const accountId = await authProvider.accountId();
    expect(accountId).toMatchSnapshot();
  });
});

describe("authenticate", () => {
  test("testnet", async () => {
    const addresses = await testnetProvider.getAccounts();
    const authProvider = new FilecoinAuthProvider(
      testnetProvider,
      addresses[0]
    );
    await expect(authProvider.authenticate("msg")).resolves.toMatchSnapshot();
  });

  test("mainnet", async () => {
    const addresses = await mainnetProvider.getAccounts();
    const authProvider = new FilecoinAuthProvider(
      mainnetProvider,
      addresses[0]
    );
    await expect(authProvider.authenticate("msg")).resolves.toMatchSnapshot();
  });

  test("mainnet BLS", async () => {
    const addresses = await blsMainnetProvider.getAccounts();
    const authProvider = new FilecoinAuthProvider(
      blsMainnetProvider,
      addresses[0]
    );
    await expect(authProvider.authenticate("msg")).resolves.toMatchSnapshot();
  });

  test("address mismatch", async () => {
    const addresses = await blsMainnetProvider.getAccounts();
    const authProvider = new FilecoinAuthProvider(
      mainnetProvider,
      addresses[0]
    );
    await expect(authProvider.authenticate("msg")).rejects.toThrow();
  });
});

describe("createLink", () => {
  test("generate proof on testnet", async () => {
    const addresses = await testnetProvider.getAccounts();
    const authProvider = new FilecoinAuthProvider(
      testnetProvider,
      addresses[0]
    );
    const proof = await authProvider.createLink(did);
    expect(proof).toMatchSnapshot();
  });

  test("generate proof on mainnet", async () => {
    const addresses = await mainnetProvider.getAccounts();
    const authProvider = new FilecoinAuthProvider(
      mainnetProvider,
      addresses[0]
    );
    const proof = await authProvider.createLink(did);
    expect(proof).toMatchSnapshot();
  });

  test("generate proof on mainnet for BLS key", async () => {
    const addresses = await blsMainnetProvider.getAccounts();
    const authProvider = new FilecoinAuthProvider(
      blsMainnetProvider,
      addresses[0]
    );
    const proof = await authProvider.createLink(did);
    expect(proof).toMatchSnapshot();
  });

  test("fail on mainnet address for testnet provider", async () => {
    const addresses = await mainnetProvider.getAccounts();
    const authProvider = new FilecoinAuthProvider(
      testnetProvider,
      addresses[0]
    );
    await expect(authProvider.createLink(did)).rejects.toThrow();
  });
  test("fail on testnet address for mainnet provider", async () => {
    const addresses = await testnetProvider.getAccounts();
    const authProvider = new FilecoinAuthProvider(
      mainnetProvider,
      addresses[0]
    );
    await expect(authProvider.createLink(did)).rejects.toThrow();
  });
});

test("withAddress", async () => {
  const addresses = await testnetProvider.getAccounts();
  const authProvider = new FilecoinAuthProvider(mainnetProvider, addresses[0]);
  const currentAccount = await authProvider.accountId();
  expect(currentAccount).toMatchSnapshot();
  const nextProvider = authProvider.withAddress("f123");
  const nextAccount = await nextProvider.accountId();
  expect(nextAccount).toMatchSnapshot();
  expect(nextAccount).not.toEqual(currentAccount);
});
