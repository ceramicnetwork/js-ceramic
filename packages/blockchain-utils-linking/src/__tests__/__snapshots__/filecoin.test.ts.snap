// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`accountId handle bls mainnet 1`] = `
AccountID {
  "address": "f3vbrwhphivdxyvs3pxpp54w73664bgxmb7ed4du4ohhu3dc6f5y264cs72yluw7mjbwnlrtzq543ys57plzka",
  "chainId": ChainID {
    "namespace": "fil",
    "reference": "f",
  },
}
`;

exports[`accountId handle mainnet 1`] = `
AccountID {
  "address": "f1ymbkm43ysdiruibji6c6m2iy5bxwbuv5d6cdqmy",
  "chainId": ChainID {
    "namespace": "fil",
    "reference": "f",
  },
}
`;

exports[`accountId handle testnet 1`] = `
AccountID {
  "address": "t17lxg2i2otnl7mmpw2ocd6o4e3b4un3272vny6ka",
  "chainId": ChainID {
    "namespace": "fil",
    "reference": "t",
  },
}
`;

exports[`authenticate mainnet 1`] = `"F5gy3IHoKUNJOUEQQMHg1iTZGa4Ef3TeT+deaBKkkYFK3Aur5y3XuOL4NNygsn9lYPnPsvXUvtBiyy4+yZ6geQE="`;

exports[`authenticate mainnet BLS 1`] = `"r1QYs+G633VEW/i5EUZVMm/SqdH5fvsylvql0chOeMJAEB6jsD9NKfgeap6mVepVE+IA2iqFdqi+F0lTkRicrxvOLs7siyPXAQZPivx34oxSEdN55LNbXL6EAIE+3p4W"`;

exports[`authenticate testnet 1`] = `"APHUqaCLuyqunqfYJl8YAq3/p6oO9iiEAGkGQILzZAEMR/L28/F+iN2ismhMJCm9yr8VjIqgneiaFIkr65fimgA="`;

exports[`createLink generate proof on mainnet 1`] = `
Object {
  "account": "f1ymbkm43ysdiruibji6c6m2iy5bxwbuv5d6cdqmy@fil:f",
  "message": "Create a new account link to your identity.

did:3:bafysdfwefwe 
Timestamp: 0",
  "signature": "V39yp3WzEVGJ0mZ5mKywD7N4jTvVhiYDAT5iGM1MjjFDeKA38hWlrMcWZx7Bz6YaYpV2GZqZjpdC64grGxMliAA=",
  "timestamp": 0,
  "type": "eoa-tx",
  "version": 2,
}
`;

exports[`createLink generate proof on mainnet for BLS key 1`] = `
Object {
  "account": "f3vbrwhphivdxyvs3pxpp54w73664bgxmb7ed4du4ohhu3dc6f5y264cs72yluw7mjbwnlrtzq543ys57plzka@fil:f",
  "message": "Create a new account link to your identity.

did:3:bafysdfwefwe 
Timestamp: 0",
  "signature": "iyq0tKJ98dHb3zMA1ijHySlEOxq9P+2dQpL6jJna8gTKRifjj0FVAsUJc5uUePC/DOw2TJA7YabYV+dQBld4BH3QYuV4Y9NNWBJCfGo8d8raKYNGxPzhSZ+5sVeawf0p",
  "timestamp": 0,
  "type": "eoa-tx",
  "version": 2,
}
`;

exports[`createLink generate proof on testnet 1`] = `
Object {
  "account": "t17lxg2i2otnl7mmpw2ocd6o4e3b4un3272vny6ka@fil:t",
  "message": "Create a new account link to your identity.

did:3:bafysdfwefwe 
Timestamp: 0",
  "signature": "UhcbW66J4k9r4GtLpyuyEMpdMydKOo+LtC5+IaM1PTYqRtACwDj86EyZDgmI7gbLPpGNyA70hwcgxIkg+nF9zgA=",
  "timestamp": 0,
  "type": "eoa-tx",
  "version": 2,
}
`;

exports[`withAddress 1`] = `
AccountID {
  "address": "t17lxg2i2otnl7mmpw2ocd6o4e3b4un3272vny6ka",
  "chainId": ChainID {
    "namespace": "fil",
    "reference": "t",
  },
}
`;

exports[`withAddress 2`] = `
AccountID {
  "address": "f123",
  "chainId": ChainID {
    "namespace": "fil",
    "reference": "f",
  },
}
`;
